front/components/assistant/conversation/AgentMessage.tsx

```diff
@@ -159,69 +159,92 @@ export function AgentMessage({
+    // understandable directly to them)
+    if (agentMessage.status === "failed") {
+      return (
+        <ErrorMessage
+          error={
+            agentMessage.error || {
+              message: "Unexpected Error",
+              code: "unexpected_error",
+            }
           }
-        }
-      />
-    );
-  }
+          retryHandler={async () => await retryHandler(agentMessage)}
+        />
+      );
+    }
 
-  // Loading state (no action nor text yet)
-  if (
-    agentMessage.status === "created" &&
-    !agentMessage.action &&
-    (!agentMessage.content || agentMessage.content === "")
-  ) {
-    return (
-      <div>
-        <div className="pb-2 text-xs font-bold text-element-600">
-          I'm thinking...
+    // Loading state (no action nor text yet)
+    if (
+      agentMessage.status === "created" &&
+      !agentMessage.action &&
+      (!agentMessage.content || agentMessage.content === "")
+    ) {
+      return (
+        <div>
+          <div className="pb-2 text-xs font-bold text-element-600">
+            I'm thinking...
+          </div>
+          <Spinner size="sm" />
         </div>
-        <Spinner size="sm" />
-      </div>
-    );
-  }
+      );
+    }
 
-  // Messages with no action and text
-  if (agentMessage.action === null && agentMessage.content) {
-    return <RenderMarkdown content={agentMessage.content} />;
+    // Messages with no action and text
+    if (agentMessage.action === null && agentMessage.content) {
+      return <RenderMarkdown content={agentMessage.content} />;
+    }
+    // Messages with action
+    if (agentMessage.action) {
+      return (
+        <>
+          <div
+            className={
+              agentMessage.content && agentMessage.content !== ""
+                ? "border-b border-dashed border-structure-300"
+                : ""
+            }
+          >
+            <AgentAction action={agentMessage.action} />
+          </div>
+          {agentMessage.content && agentMessage.content !== "" && (
+            <>
+              <div className="pt-4">
+                <RenderMarkdown content={agentMessage.content} />
+              </div>
+            </>
+          )}
+        </>
+      );
+    }
   }
-  // Messages with action
-  if (agentMessage.action) {
-    return (
-      <>
-        <div
-          className={
-            agentMessage.content && agentMessage.content !== ""
-              ? "border-b border-dashed border-structure-300"
-              : ""
-          }
-        >
-          <AgentAction action={agentMessage.action} />
-        </div>
-        {agentMessage.content && agentMessage.content !== "" && (
-          <>
-            <div className="pt-4">
-              <RenderMarkdown content={agentMessage.content} />
-            </div>
-          </>
-        )}
-      </>
+
+  async function retryHandler(agentMessage: AgentMessageType) {
+    await fetch(
+      `/api/w/${owner.sId}/assistant/conversations/${conversationId}/messages/retry`,
```

Comment:
URL should be `/api/w/${owner.sId}/assistant/conversations/${conversationId}/messages/[mId]/retry`

Reply:
I had initially done that, but since the retried message will not have the same sId as the original I moved it
It felt more similar to the POST new message which is on [cid]/messages/index.tsx (creation of a new message with a new sId)

Not strongly held, the mId route is fine too and I can change if you prefer;  just wanted to double-check

Reply:
Ok, will change